File: backend/services/ldap.py
from __future__ import annotations

import asyncio
from datetime import UTC, datetime
from typing import Any

from profiles import HoneypotProfile
from services.base import BaseHoneypotService


class LDAPHoneypot(BaseHoneypotService):
    def __init__(
        self,
        name: str,
        host: str,
        port: int,
        logger,
        profile: HoneypotProfile,
    ) -> None:
        super().__init__(name=name, host=host, port=port, logger=logger)
        self.profile = profile

    def set_profile(self, profile: HoneypotProfile) -> None:
        self.profile = profile

    async def handle_client(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ) -> None:
        src_ip, src_port = self.peer(writer)
        await self.log_event("connection", src_ip=src_ip, src_port=src_port)
        try:
            while True:
                packet = await asyncio.wait_for(reader.read(4096), timeout=15.0)
                if not packet:
                    break
                handled = await self._handle_packet(packet, src_ip=src_ip, src_port=src_port, writer=writer)
                if not handled:
                    break
        except (BrokenPipeError, ConnectionResetError):
            await self.log_event("client_disconnected", src_ip=src_ip, src_port=src_port)
        except Exception as exc:
            await self.log_event(
                "connection_error",
                src_ip=src_ip,
                src_port=src_port,
                error=type(exc).__name__,
            )
        finally:
            await self.close_writer(writer)

    async def _handle_packet(
        self,
        packet: bytes,
        *,
        src_ip: str,
        src_port: int,
        writer: asyncio.StreamWriter,
    ) -> bool:
        message_id, protocol_op, payload = _parse_ldap_message(packet)
        if protocol_op == 0x60:
            username, password = _parse_bind_request(payload)
            await self.log_event(
                "login_attempt",
                src_ip=src_ip,
                src_port=src_port,
                username=username,
                password=password,
                summary=f"LDAP bind attempt for {username or '<anonymous>'}",
            )
            await self.write_bytes(
                writer,
                _wrap_ldap_message(
                    message_id,
                    bytes([0x61]) + _ber_length(len(_build_bind_response_payload())) + _build_bind_response_payload(),
                ),
            )
            return True

        if protocol_op == 0x63:
            base_dn, scope = _parse_search_request(payload)
            await self.log_event(
                "ldap_search",
                src_ip=src_ip,
                src_port=src_port,
                base_dn=base_dn,
                scope=scope,
                summary=f"LDAP search under {base_dn or '<rootDSE>'}",
            )
            
            responses = []
            if not base_dn:
                responses.extend(_build_rootdse_search_response(message_id, self.profile))
            else:
                mock_entries = _build_ad_mock_entries(self.profile)
                for entry in mock_entries:
                    responses.append(_build_search_entry_response(message_id, entry["dn"], entry["attributes"]))
                
                done_payload = _build_generic_result_payload(result_code=0, diagnostic_message="")
                search_done = _wrap_ldap_message(
                    message_id,
                    bytes([0x65]) + _ber_length(len(done_payload)) + done_payload,
                )
                responses.append(search_done)
                
            for response_packet in responses:
                await self.write_bytes(writer, response_packet)
            return True

        if protocol_op == 0x42:
            await self.log_event(
                "ldap_unbind",
                src_ip=src_ip,
                src_port=src_port,
                summary="LDAP unbind received.",
            )
            return False

        await self.log_event(
            "ldap_operation",
            src_ip=src_ip,
            src_port=src_port,
            operation=f"0x{protocol_op:02x}",
            summary=f"Unhandled LDAP op 0x{protocol_op:02x}",
        )
        response_payload = _build_generic_result_payload(result_code=2, diagnostic_message="Protocol error")
        await self.write_bytes(
            writer,
            _wrap_ldap_message(
                message_id,
                bytes([0x65]) + _ber_length(len(response_payload)) + response_payload,
            ),
        )
        return False

def _parse_ldap_message(packet: bytes) -> tuple[int, int, bytes]:
    tag, _, body, _ = _read_tlv(packet, 0)
    if tag != 0x30:
        raise ValueError("LDAP packet is not a sequence.")
    _, _, message_id_bytes, offset = _read_tlv(body, 0)
    _, _, protocol_payload, _ = _read_tlv(body, offset)
    protocol_op = body[offset]
    return int.from_bytes(message_id_bytes, "big"), protocol_op, protocol_payload


def _parse_bind_request(payload: bytes) -> tuple[str, str]:
    _, _, _, offset = _read_tlv(payload, 0)
    _, _, name_bytes, offset = _read_tlv(payload, offset)
    username = name_bytes.decode("utf-8", errors="replace")
    password = ""
    if offset < len(payload):
        auth_tag = payload[offset]
        _, _, auth_value, _ = _read_tlv(payload, offset)
        if auth_tag == 0x80:
            password = auth_value.decode("utf-8", errors="replace")
    return username, password


def _parse_search_request(payload: bytes) -> tuple[str, str]:
    _, _, base_dn_bytes, offset = _read_tlv(payload, 0)
    _, _, scope_bytes, _ = _read_tlv(payload, offset)
    base_dn = base_dn_bytes.decode("utf-8", errors="replace")
    scope = {
        b"\x00": "baseObject",
        b"\x01": "singleLevel",
        b"\x02": "wholeSubtree",
    }.get(scope_bytes, scope_bytes.hex())
    return base_dn, scope


def _build_bind_response_payload() -> bytes:
    return _build_generic_result_payload(
        result_code=49,
        diagnostic_message=(
            "80090308: LdapErr: DSID-0C090457, comment: AcceptSecurityContext error, "
            "data 52e, v4563"
        ),
    )


def _build_generic_result_payload(*, result_code: int, diagnostic_message: str) -> bytes:
    return b"".join(
        [
            _ber_enumerated(result_code),
            _ber_octet_string(""),
            _ber_octet_string(diagnostic_message),
        ]
    )


def _build_rootdse_search_response(message_id: int, profile: HoneypotProfile) -> list[bytes]:
    current_time = datetime.now(UTC).strftime("%Y%m%d%H%M%S.0Z")
    dns_domain = profile.smb.dns_domain
    dc_suffix = ",".join(f"DC={part}" for part in dns_domain.split("."))
    attributes = [
        ("currentTime", [current_time]),
        ("defaultNamingContext", [dc_suffix]),
        ("dnsHostName", [f"{profile.smb.hostname.upper()}.{dns_domain}"]),
        ("supportedLDAPVersion", ["3"]),
    ]
    entry_payload = b"".join(
        [
            _ber_octet_string(""),
            _ber_sequence(b"".join(_build_partial_attribute(name, values) for name, values in attributes)),
        ]
    )
    search_entry = _wrap_ldap_message(
        message_id,
        bytes([0x64]) + _ber_length(len(entry_payload)) + entry_payload,
    )
    done_payload = _build_generic_result_payload(result_code=0, diagnostic_message="")
    search_done = _wrap_ldap_message(
        message_id,
        bytes([0x65]) + _ber_length(len(done_payload)) + done_payload,
    )
    return [search_entry, search_done]


def _build_partial_attribute(name: str, values: list[str]) -> bytes:
    value_set = bytes([0x31]) + _ber_length(
        sum(len(_ber_octet_string(value)) for value in values)
    ) + b"".join(_ber_octet_string(value) for value in values)
    return _ber_sequence(_ber_octet_string(name) + value_set)


def _wrap_ldap_message(message_id: int, protocol_op: bytes) -> bytes:
    body = _ber_integer(message_id) + protocol_op
    return _ber_sequence(body)


def _ber_sequence(payload: bytes) -> bytes:
    return bytes([0x30]) + _ber_length(len(payload)) + payload


def _ber_integer(value: int) -> bytes:
    if value == 0:
        encoded = b"\x00"
    else:
        encoded = value.to_bytes((value.bit_length() + 7) // 8, "big")
        if encoded[0] & 0x80:
            encoded = b"\x00" + encoded
    return bytes([0x02]) + _ber_length(len(encoded)) + encoded


def _ber_enumerated(value: int) -> bytes:
    encoded = value.to_bytes(1, "big")
    return bytes([0x0A]) + _ber_length(len(encoded)) + encoded


def _ber_octet_string(value: str) -> bytes:
    encoded = value.encode("utf-8")
    return bytes([0x04]) + _ber_length(len(encoded)) + encoded


def _ber_length(length: int) -> bytes:
    if length < 0x80:
        return bytes([length])
    encoded = length.to_bytes((length.bit_length() + 7) // 8, "big")
    return bytes([0x80 | len(encoded)]) + encoded


def _read_tlv(data: bytes, offset: int) -> tuple[int, int, bytes, int]:
    if offset >= len(data):
        raise ValueError("Missing BER tag.")
    tag = data[offset]
    offset += 1
    if offset >= len(data):
        raise ValueError("Missing BER length.")
    first_length = data[offset]
    offset += 1
    if first_length & 0x80:
        size = first_length & 0x7F
        length = int.from_bytes(data[offset : offset + size], "big")
        offset += size
    else:
        length = first_length
    end = offset + length
    if end > len(data):
        raise ValueError("Truncated BER payload.")
    return tag, length, data[offset:end], end


def _build_search_entry_response(message_id: int, dn: str, attributes: list[tuple[str, list[str]]]) -> bytes:
    entry_payload = b"".join(
        [
            _ber_octet_string(dn),
            _ber_sequence(b"".join(_build_partial_attribute(name, values) for name, values in attributes)),
        ]
    )
    return _wrap_ldap_message(
        message_id,
        bytes([0x64]) + _ber_length(len(entry_payload)) + entry_payload,
    )


def _build_ad_mock_entries(profile: HoneypotProfile) -> list[dict[str, Any]]:
    dns_domain = profile.smb.dns_domain
    dc_suffix = ",".join(f"DC={part}" for part in dns_domain.split("."))
    
    return [
        {
            "dn": f"CN=Administrator,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "person", "organizationalPerson", "user"]),
                ("cn", ["Administrator"]),
                ("sAMAccountName", ["Administrator"]),
                ("userPrincipalName", [f"Administrator@{dns_domain}"]),
                ("memberOf", [f"CN=Domain Admins,CN=Users,{dc_suffix}", f"CN=Enterprise Admins,CN=Users,{dc_suffix}"]),
                ("pwdLastSet", ["132649032000000000"]),
                ("userAccountControl", ["512"]),
            ]
        },
        {
            "dn": f"CN=cagri,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "person", "organizationalPerson", "user"]),
                ("cn", ["cagri"]),
                ("sAMAccountName", ["cagri"]),
                ("userPrincipalName", [f"cagri@{dns_domain}"]),
                ("memberOf", [f"CN=Domain Users,CN=Users,{dc_suffix}"]),
                ("pwdLastSet", ["132849032000000000"]),
                ("userAccountControl", ["512"]),
            ]
        },
        {
            "dn": f"CN=sql_service,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "person", "organizationalPerson", "user"]),
                ("cn", ["sql_service"]),
                ("sAMAccountName", ["sql_service"]),
                ("userPrincipalName", [f"sql_service@{dns_domain}"]),
                ("memberOf", [f"CN=Domain Users,CN=Users,{dc_suffix}"]),
                ("pwdLastSet", ["132949032000000000"]),
                ("servicePrincipalName", [f"MSSQLSvc/{profile.smb.hostname.upper()}.{dns_domain}:1433"]),
                ("userAccountControl", ["512"]),
            ]
        },
        {
            "dn": f"CN={profile.smb.hostname.upper()},OU=Domain Controllers,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "person", "organizationalPerson", "user", "computer"]),
                ("cn", [profile.smb.hostname.upper()]),
                ("sAMAccountName", [f"{profile.smb.hostname.upper()}$"]),
                ("dnsHostName", [f"{profile.smb.hostname.upper()}.{dns_domain}"]),
                ("userAccountControl", ["532480"]),
            ]
        },
        {
            "dn": f"CN=SQL-PROD-01,CN=Computers,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "person", "organizationalPerson", "user", "computer"]),
                ("cn", ["SQL-PROD-01"]),
                ("sAMAccountName", ["SQL-PROD-01$"]),
                ("dnsHostName", [f"SQL-PROD-01.{dns_domain}"]),
                ("userAccountControl", ["4096"]),
            ]
        },
        {
            "dn": f"CN=Domain Admins,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "group"]),
                ("cn", ["Domain Admins"]),
                ("sAMAccountName", ["Domain Admins"]),
                ("member", [f"CN=Administrator,CN=Users,{dc_suffix}"]),
            ]
        },
        {
            "dn": f"CN=Enterprise Admins,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "group"]),
                ("cn", ["Enterprise Admins"]),
                ("sAMAccountName", ["Enterprise Admins"]),
                ("member", [f"CN=Administrator,CN=Users,{dc_suffix}"]),
            ]
        },
        {
            "dn": f"CN=Domain Users,CN=Users,{dc_suffix}",
            "attributes": [
                ("objectClass", ["top", "group"]),
                ("cn", ["Domain Users"]),
                ("sAMAccountName", ["Domain Users"]),
                ("member", [f"CN=Administrator,CN=Users,{dc_suffix}", f"CN=cagri,CN=Users,{dc_suffix}", f"CN=sql_service,CN=Users,{dc_suffix}"]),
            ]
        }
    ]

File: backend/internal/services/tcp/ldap.go
package tcp

import (
	"context"
	"fmt"
	"net"
	"strings"
	"time"

	"honeypot-orchestrator/backend/internal/defense"
	"honeypot-orchestrator/backend/internal/services"
	"honeypot-orchestrator/backend/internal/logger"
	"honeypot-orchestrator/backend/internal/profiles"
)

type LDAPHoneypot struct {
	baseService *services.BaseTCPService
	profile     *profiles.HoneypotProfile
}

func NewLDAPHoneypot(
	name string,
	host string,
	port int,
	el *logger.EventLogger,
	ds *defense.DefenseSystem,
	initialProfile *profiles.HoneypotProfile,
) *LDAPHoneypot {
	l := &LDAPHoneypot{profile: initialProfile}
	l.baseService = services.NewBaseTCPService(name, host, port, el, ds, l.handleClient)
	return l
}

func (l *LDAPHoneypot) Name() string {
	return l.baseService.Name()
}

func (l *LDAPHoneypot) Port() int {
	return l.baseService.Port()
}

func (l *LDAPHoneypot) IsRunning() bool {
	return l.baseService.IsRunning()
}

func (l *LDAPHoneypot) Start(ctx context.Context) error {
	return l.baseService.Start(ctx)
}

func (l *LDAPHoneypot) Stop() error {
	return l.baseService.Stop()
}

func (l *LDAPHoneypot) Proto() string {
	return l.baseService.Proto()
}

func (l *LDAPHoneypot) SetProfile(prof *profiles.HoneypotProfile) {
	l.profile = prof
}

func (l *LDAPHoneypot) handleClient(ctx context.Context, conn net.Conn) error {
	remoteAddr := conn.RemoteAddr()
	tcpAddr, ok := remoteAddr.(*net.TCPAddr)
	if !ok {
		return fmt.Errorf("remote address is not TCP")
	}
	srcIP := tcpAddr.IP.String()
	srcPort := tcpAddr.Port

	l.baseService.LogEvent("connection", map[string]interface{}{
		"src_ip":   srcIP,
		"src_port": srcPort,
	})

	for {
		conn.SetReadDeadline(time.Now().Add(15 * time.Second))
		buf := make([]byte, 4096)
		n, err := conn.Read(buf)
		if err != nil || n == 0 {
			break
		}

		packet := buf[:n]
		handled, err := l.handlePacket(conn, packet, srcIP, srcPort)
		if err != nil || !handled {
			break
		}
	}

	return nil
}

func (l *LDAPHoneypot) handlePacket(conn net.Conn, packet []byte, srcIP string, srcPort int) (bool, error) {
	messageID, protocolOp, payload, err := parseLdapMessage(packet)
	if err != nil {
		return false, err
	}

	conn.SetWriteDeadline(time.Now().Add(5 * time.Second))

	if protocolOp == 0x60 { // Bind Request
		username, password := parseBindRequest(payload)
		l.baseService.LogEvent("login_attempt", map[string]interface{}{
			"src_ip":   srcIP,
			"src_port": srcPort,
			"username": username,
			"password": password,
			"summary":  fmt.Sprintf("LDAP bind attempt for %s", username),
		})

		bindResp := buildBindResponsePayload()
		resp := wrapLdapMessage(messageID, append([]byte{0x61}, append(berLength(len(bindResp)), bindResp...)...))
		_, err = conn.Write(resp)
		return true, err
	}

	if protocolOp == 0x63 { // Search Request
		baseDN, scope := parseSearchRequest(payload)
		l.baseService.LogEvent("ldap_search", map[string]interface{}{
			"src_ip":   srcIP,
			"src_port": srcPort,
			"base_dn":  baseDN,
			"scope":    scope,
			"summary":  fmt.Sprintf("LDAP search under %s", baseDN),
		})

		var responses [][]byte
		if baseDN == "" { // rootDSE
			responses = buildRootDSESearchResponse(messageID, l.profile)
		} else {
			mockEntries := buildADMockEntries(l.profile)
			for _, entry := range mockEntries {
				responses = append(responses, buildSearchEntryResponse(messageID, entry.DN, entry.Attributes))
			}
			donePayload := buildGenericResultPayload(0, "")
			searchDone := wrapLdapMessage(messageID, append([]byte{0x65}, append(berLength(len(donePayload)), donePayload...)...))
			responses = append(responses, searchDone)
		}

		for _, respPacket := range responses {
			if _, err := conn.Write(respPacket); err != nil {
				return false, err
			}
		}
		return true, nil
	}

	if protocolOp == 0x42 { // Unbind Request
		l.baseService.LogEvent("ldap_unbind", map[string]interface{}{
			"src_ip":   srcIP,
			"src_port": srcPort,
			"summary":  "LDAP unbind received.",
		})
		return false, nil
	}

	l.baseService.LogEvent("ldap_operation", map[string]interface{}{
		"src_ip":    srcIP,
		"src_port":  srcPort,
		"operation": fmt.Sprintf("0x%02x", protocolOp),
		"summary":   fmt.Sprintf("Unhandled LDAP op 0x%02x", protocolOp),
	})

	errPayload := buildGenericResultPayload(2, "Protocol error")
	resp := wrapLdapMessage(messageID, append([]byte{0x65}, append(berLength(len(errPayload)), errPayload...)...))
	_, _ = conn.Write(resp)
	return false, nil
}

func readTLV(data []byte, offset int) (byte, int, []byte, int, error) {
	if offset >= len(data) {
		return 0, 0, nil, 0, fmt.Errorf("missing BER tag")
	}
	tag := data[offset]
	offset++
	if offset >= len(data) {
		return 0, 0, nil, 0, fmt.Errorf("missing BER length")
	}
	firstLength := data[offset]
	offset++

	var length int
	if firstLength&0x80 != 0 {
		size := int(firstLength & 0x7F)
		if offset+size > len(data) {
			return 0, 0, nil, 0, fmt.Errorf("truncated BER length size")
		}
		val := 0
		for i := 0; i < size; i++ {
			val = (val << 8) | int(data[offset+i])
		}
		length = val
		offset += size
	} else {
		length = int(firstLength)
	}

	end := offset + length
	if end > len(data) {
		return 0, 0, nil, 0, fmt.Errorf("truncated BER payload")
	}

	return tag, length, data[offset:end], end, nil
}

func parseLdapMessage(packet []byte) (int, byte, []byte, error) {
	tag, _, body, _, err := readTLV(packet, 0)
	if err != nil {
		return 0, 0, nil, err
	}
	if tag != 0x30 {
		return 0, 0, nil, fmt.Errorf("LDAP packet is not a sequence")
	}

	idTag, _, idBytes, offset, err := readTLV(body, 0)
	if err != nil || idTag != 0x02 {
		return 0, 0, nil, fmt.Errorf("invalid message ID tag")
	}

	messageID := 0
	for _, b := range idBytes {
		messageID = (messageID << 8) | int(b)
	}

	if offset >= len(body) {
		return 0, 0, nil, fmt.Errorf("missing protocol payload")
	}
	protocolOp := body[offset]
	_, _, protocolPayload, _, err := readTLV(body, offset)
	if err != nil {
		return 0, 0, nil, err
	}

	return messageID, protocolOp, protocolPayload, nil
}

func parseBindRequest(payload []byte) (string, string) {
	versionTag, _, _, offset, err := readTLV(payload, 0)
	if err != nil || versionTag != 0x02 {
		return "", ""
	}
	nameTag, _, nameBytes, offset, err := readTLV(payload, offset)
	if err != nil || nameTag != 0x04 {
		return "", ""
	}
	username := string(nameBytes)
	password := ""

	if offset < len(payload) {
		authTag := payload[offset]
		_, _, authValue, _, err := readTLV(payload, offset)
		if err == nil && authTag == 0x80 {
			password = string(authValue)
		}
	}
	return username, password
}

func parseSearchRequest(payload []byte) (string, string) {
	baseTag, _, baseBytes, offset, err := readTLV(payload, 0)
	if err != nil || baseTag != 0x04 {
		return "", ""
	}
	baseDN := string(baseBytes)

	scopeTag, _, scopeBytes, _, err := readTLV(payload, offset)
	if err != nil || scopeTag != 0x0a {
		return baseDN, "unknown"
	}

	scope := "unknown"
	if len(scopeBytes) == 1 {
		switch scopeBytes[0] {
		case 0x00:
			scope = "baseObject"
		case 0x01:
			scope = "singleLevel"
		case 0x02:
			scope = "wholeSubtree"
		}
	}
	return baseDN, scope
}

func buildBindResponsePayload() []byte {
	return buildGenericResultPayload(49, "80090308: LdapErr: DSID-0C090457, comment: AcceptSecurityContext error, data 52e, v4563")
}

func buildGenericResultPayload(resultCode int, diagMsg string) []byte {
	var payload []byte
	payload = append(payload, berEnumerated(resultCode)...)
	payload = append(payload, berOctetString("")...)
	payload = append(payload, berOctetString(diagMsg)...)
	return payload
}

func buildRootDSESearchResponse(messageID int, profile *profiles.HoneypotProfile) [][]byte {
	currentTime := time.Now().UTC().Format("20060102150405.0Z")
	dnsDomain := profile.SMB.DNSDomain
	var dcParts []string
	for _, part := range strings.Split(dnsDomain, ".") {
		dcParts = append(dcParts, "DC="+part)
	}
	dcSuffix := strings.Join(dcParts, ",")

	attrs := []struct {
		Name   string
		Values []string
	}{
		{"currentTime", []string{currentTime}},
		{"defaultNamingContext", []string{dcSuffix}},
		{"dnsHostName", []string{fmt.Sprintf("%s.%s", strings.ToUpper(profile.SMB.Hostname), dnsDomain)}},
		{"supportedLDAPVersion", []string{"3"}},
	}

	var entryPayload []byte
	entryPayload = append(entryPayload, berOctetString("")...)
	var partials []byte
	for _, attr := range attrs {
		partials = append(partials, buildPartialAttribute(attr.Name, attr.Values)...)
	}
	entryPayload = append(entryPayload, berSequence(partials)...)

	searchEntry := wrapLdapMessage(messageID, append([]byte{0x64}, append(berLength(len(entryPayload)), entryPayload...)...))
	donePayload := buildGenericResultPayload(0, "")
	searchDone := wrapLdapMessage(messageID, append([]byte{0x65}, append(berLength(len(donePayload)), donePayload...)...))

	return [][]byte{searchEntry, searchDone}
}

func buildPartialAttribute(name string, values []string) []byte {
	var valSetBytes []byte
	for _, val := range values {
		valSetBytes = append(valSetBytes, berOctetString(val)...)
	}
	valueSet := append([]byte{0x31}, append(berLength(len(valSetBytes)), valSetBytes...)...)
	return berSequence(append(berOctetString(name), valueSet...))
}

func wrapLdapMessage(messageID int, protocolOp []byte) []byte {
	body := append(berInteger(messageID), protocolOp...)
	return berSequence(body)
}

func berSequence(payload []byte) []byte {
	return append([]byte{0x30}, append(berLength(len(payload)), payload...)...)
}

func berInteger(val int) []byte {
	var encoded []byte
	if val == 0 {
		encoded = []byte{0x00}
	} else {
		temp := val
		for temp > 0 {
			encoded = append([]byte{byte(temp & 0xFF)}, encoded...)
			temp = temp >> 8
		}
		if encoded[0]&0x80 != 0 {
			encoded = append([]byte{0x00}, encoded...)
		}
	}
	return append([]byte{0x02}, append(berLength(len(encoded)), encoded...)...)
}

func berEnumerated(val int) []byte {
	return []byte{0x0A, 0x01, byte(val)}
}

func berOctetString(val string) []byte {
	encoded := []byte(val)
	return append([]byte{0x04}, append(berLength(len(encoded)), encoded...)...)
}

func berLength(length int) []byte {
	if length < 0x80 {
		return []byte{byte(length)}
	}
	var encoded []byte
	temp := length
	for temp > 0 {
		encoded = append([]byte{byte(temp & 0xFF)}, encoded...)
		temp = temp >> 8
	}
	return append([]byte{0x80 | byte(len(encoded))}, encoded...)
}

type ldapMockEntry struct {
	DN         string
	Attributes []struct {
		Name   string
		Values []string
	}
}

func buildADMockEntries(profile *profiles.HoneypotProfile) []ldapMockEntry {
	dnsDomain := profile.SMB.DNSDomain
	var dcParts []string
	for _, part := range strings.Split(dnsDomain, ".") {
		dcParts = append(dcParts, "DC="+part)
	}
	dcSuffix := strings.Join(dcParts, ",")

	makeAttr := func(name string, values []string) struct {
		Name   string
		Values []string
	} {
		return struct {
			Name   string
			Values []string
		}{Name: name, Values: values}
	}

	return []ldapMockEntry{
		{
			DN: fmt.Sprintf("CN=Administrator,CN=Users,%s", dcSuffix),
			Attributes: []struct {
				Name   string
				Values []string
			}{
				makeAttr("objectClass", []string{"top", "person", "organizationalPerson", "user"}),
				makeAttr("cn", []string{"Administrator"}),
				makeAttr("sAMAccountName", []string{"Administrator"}),
				makeAttr("userPrincipalName", []string{fmt.Sprintf("Administrator@%s", dnsDomain)}),
				makeAttr("memberOf", []string{fmt.Sprintf("CN=Domain Admins,CN=Users,%s", dcSuffix), fmt.Sprintf("CN=Enterprise Admins,CN=Users,%s", dcSuffix)}),
				makeAttr("pwdLastSet", []string{"132649032000000000"}),
				makeAttr("userAccountControl", []string{"512"}),
			},
		},
		{
			DN: fmt.Sprintf("CN=cagri,CN=Users,%s", dcSuffix),
			Attributes: []struct {
				Name   string
				Values []string
			}{
				makeAttr("objectClass", []string{"top", "person", "organizationalPerson", "user"}),
				makeAttr("cn", []string{"cagri"}),
				makeAttr("sAMAccountName", []string{"cagri"}),
				makeAttr("userPrincipalName", []string{fmt.Sprintf("cagri@%s", dnsDomain)}),
				makeAttr("memberOf", []string{fmt.Sprintf("CN=Domain Users,CN=Users,%s", dcSuffix)}),
				makeAttr("pwdLastSet", []string{"132849032000000000"}),
				makeAttr("userAccountControl", []string{"512"}),
			},
		},
		{
			DN: fmt.Sprintf("CN=sql_service,CN=Users,%s", dcSuffix),
			Attributes: []struct {
				Name   string
				Values []string
			}{
				makeAttr("objectClass", []string{"top", "person", "organizationalPerson", "user"}),
				makeAttr("cn", []string{"sql_service"}),
				makeAttr("sAMAccountName", []string{"sql_service"}),
				makeAttr("userPrincipalName", []string{fmt.Sprintf("sql_service@%s", dnsDomain)}),
				makeAttr("memberOf", []string{fmt.Sprintf("CN=Domain Users,CN=Users,%s", dcSuffix)}),
				makeAttr("pwdLastSet", []string{"132949032000000000"}),
				makeAttr("servicePrincipalName", []string{fmt.Sprintf("MSSQLSvc/%s.%s:1433", strings.ToUpper(profile.SMB.Hostname), dnsDomain)}),
				makeAttr("userAccountControl", []string{"512"}),
			},
		},
		{
			DN: fmt.Sprintf("CN=%s,OU=Domain Controllers,%s", strings.ToUpper(profile.SMB.Hostname), dcSuffix),
			Attributes: []struct {
				Name   string
				Values []string
			}{
				makeAttr("objectClass", []string{"top", "person", "organizationalPerson", "user", "computer"}),
				makeAttr("cn", []string{strings.ToUpper(profile.SMB.Hostname)}),
				makeAttr("sAMAccountName", []string{fmt.Sprintf("%s$", strings.ToUpper(profile.SMB.Hostname))}),
				makeAttr("dnsHostName", []string{fmt.Sprintf("%s.%s", strings.ToUpper(profile.SMB.Hostname), dnsDomain)}),
				makeAttr("userAccountControl", []string{"532480"}),
			},
		},
	}
}

func buildSearchEntryResponse(messageID int, dn string, attributes []struct {
	Name   string
	Values []string
}) []byte {
	var partials []byte
	for _, attr := range attributes {
		partials = append(partials, buildPartialAttribute(attr.Name, attr.Values)...)
	}
	entryPayload := append(berOctetString(dn), berSequence(partials)...)
	return wrapLdapMessage(messageID, append([]byte{0x64}, append(berLength(len(entryPayload)), entryPayload...)...))
}

func init() {
	services.Registry["ldap"] = func(name, host string, port int, el *logger.EventLogger, ds *defense.DefenseSystem, prof *profiles.HoneypotProfile) services.HoneypotService {
		return NewLDAPHoneypot(name, host, port, el, ds, prof)
	}
}

func (s *LDAPHoneypot) PortNameHost() string {
	return s.baseService.PortNameHost()
}
